src/windows/line.rs

```diff
@@ -0,0 +1,150 @@
+// Licensed under the Apache License, Version 2.0, <LICENSE-APACHE or
+// http://apache.org/licenses/LICENSE-2.0> or the MIT license <LICENSE-MIT or
+// http://opensource.org/licenses/MIT>, at your option. This file may not be
+// copied, modified, or distributed except according to those terms.
+
+use pdb::{AddressMap, PdbInternalRva};
+use std::fmt::{Display, Formatter};
+
+#[derive(Clone, Debug)]
+pub struct Line {
+    pub rva: u32,
+    pub num: u32,
+    pub len: u32,
+    pub file_id: u32,
+}
+
+#[derive(Debug, Default)]
+pub struct Lines {
+    pub lines: Vec<Line>,
+    pub is_sorted: bool,
+}
+
+impl Display for Lines {
+    fn fmt(&self, f: &mut Formatter) -> std::fmt::Result {
+        for line in self.lines.iter() {
+            writeln!(
+                f,
+                "{:x} {:x} {} {}",
+                line.rva, line.len, line.num, line.file_id
+            )?;
+        }
+        Ok(())
+    }
+}
+
+impl Lines {
+    fn compute_len(&mut self, sym_len: u32) {
+        // The length (in the binary) of the line is not in the pdb but we can infer it:
+        // RVA     LINE NUMBER
+        // 0x0001  10  <= the size of line 10 is 0x000B - 0x0001
+        // 0x000B  11
+        // ...
+        // 0x002A  15 <= the size of line 15 is sym length - (0x002A - 0x0001)
+
+        if self.lines.is_empty() {
+            return;
+        }
+
+        if !self.is_sorted {
+            self.lines.sort_by_key(|x| x.rva);
+        }
+
+        let n_lines = self.lines.len();
+
+        unsafe {
+            let first_rva = if n_lines > 1 {
```

Comment:
I think the logic here is somewhat obscured.  I suggest something like:

```
let first_rva = self.lines[0].rva;

// Update length of all lines.
if n_lines > 1 {
  for i in 0..n_lines - 1 {
  }
}
 
// Update the length of the last line
```

I wanted to use `.windows_mut(2)`, but apparently that method on slices doesn't exist because of unsoundness concerns. :(

Reply:
Done: I removed that stuff and replace it with a smaller and safe code.